File: CSE241/HW8/Tetris.java
/**
 * 
 * @author Burak Demirkaya 210104004274
 * 
 */

import java.util.Random;

/**
 * Tetris board class
 * 
 */
public class Tetris {

    /**
     * 2D char array to represent the tetris game.
     */
    private char[][] Board;
    /**
     * Number of rows of the board.
     */
    private int board_row;

    /**
     * Number of columns of the board.
     */
    private int board_col;


    /**
     * Creates a row x col board and fills it border with '#' so that the borders can be easily seen.
     * 
     * @param row Row number of the Tetris board
     * @param col Column number of the Tetris board
     * 
     */
    Tetris(int row,int col){
        board_row = row;
        board_col = col;
        char[][] temp = new char[row][col];

        for(int r=0; r<board_row; ++r){
            for(int c=0; c<board_col; ++c){
                if(r == 0 || r==board_row-1 || c == 0 || c == board_col-1){
                    temp[r][c] = '#';
                }
                else{
                    temp[r][c] = ' ';
                }            
            }
        }
        Board = temp;
    }

    /**
     * Adds the tetromino to the middle row of the Tetris board.
     * 
     * @param t Takes a Tetromino object
     */
    public void add(Tetromino t){
        int mid_col = board_col / 2 ;
        char[][] shape = t.getShape();
        for(int r=0; r<shape.length; ++r){
            for(int c=0; c<shape[r].length; ++c){
                Board[r+1][c+mid_col-1] = shape[r][c];
            }
        }
    }

    /**
     * Deletes the added tetromino from the middle of the Tetris board.
     * 
     * @param t Takes a tetromino object
     */
    public void deleteAdd(Tetromino t){
        int mid_col = board_col / 2;
        char[][] shape = t.getShape();
        for(int r=0; r<shape.length; ++r){
            for(int c=0; c<shape[r].length; ++c){
                Board[r+1][mid_col + c -1] = ' ';
            }
        }
    }

    /**
     * 
     * Prints the Tetris board.
     */
    public void draw(){
        for(int r=0; r<board_row; ++r){
            for(int c=0; c<board_col; ++c){
                System.out.print(Board[r][c]);
            }
            System.out.println();
        }
    }

    /**
     * First by random calculate the rotation count rotation direction to rotate the tetromino.
     * Then calculates by random the move count and move direction. If move direction is left it multiplies
     * move count by -1.
     * Finally drop downs the tetromino on the Tetris board until it hits bottom or hits another tetromino.
     * 
     * @param t Takes a tetromino object
     * @param game Takes a game object to paint the game board in GUI.
     */
    public void animate(Tetromino t, Game game){
        int bottom = board_row - 1;
        int count = 0;
        Random r = new Random();
        int rotate_count = r.nextInt(5); // Random integers in range of [0,4]
        int rotate_direction = r.nextInt(2) + 1; // Random integers in range of [1,2]

        if(rotate_direction == 1){ // Left rotate
            for(int i=0; i<rotate_count; ++i){
                t.rotate('l');
            }
        }
        else if(rotate_direction == 2){ // Right rotate
            for(int i=0; i<rotate_count; ++i){
                t.rotate('r');
            }
        }

        int move_count = r.nextInt(board_col/2-1-t.getCol());
        int move_direction = r.nextInt(2) + 1;
        if(move_direction == 1){ // Left move
            move_count *= -1;
        }

        char[][] shape = t.getShape();

        while(count <= bottom){
            if(move_direction == 'l'){
                t.moveLeft();
            }
            else if(move_direction == 'r'){
                t.moveRight();
            }
            draw();
            deleteAdd(t);
            
            for(int i=0; i<t.getCol(); ++i){ // Collision check
                if((Board[t.getRow()+count+1+i][t.getCol()+i+count] != ' ')){
                    return;
                }
            }

            for(int r1=0; r1<shape.length; ++r1){ // Erases the previous point
                for(int c=0; c<shape[0].length; ++c){
                    Board[r1+count+1][board_col/2 + move_count +c-1] = ' ';
                }
            }

            game.colorBoard(Board);

            for(int r2=0; r2<shape.length; ++r2){ // Puts shape to the board
                for(int c=0; c<shape[0].length; ++c){
                    Board[r2+count+2][board_col/2 + move_count + c-1] = shape[r2][c];
                }
            }

            game.colorBoard(Board);
                        
            sleep();
            count++;


            if(count+shape.length >= bottom-1){
                break;
            }
        }
        draw();
    }
    /**
     * Sleeps for 50 miliseconds than deletes the entire screen and flushes the output buffer.
     * And set the cursor to the top of the terminal.
     * 
     */
    public void sleep() {
        try {
            Thread.sleep(50); // Sleeps 50 milliseconds
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.print("\033[H\033[2J");   // Erases entire screen
        System.out.flush(); // flush the output buffer
    }

    /**
     *  If the board's row 2 is full, it returns true. Otherwise returns false.
     * 
     * @return boolean.
     */
    public boolean boardFull(){
        boolean full = true;
        for(int i=1; i<board_col-1; ++i){
            if(Board[2][i] == ' '){
                full = false;
                break;
            }
        }
        return full;
    }
    /**
     * Checks for columns if all full return true, otherwise return false.
     * 
     * @return boolean. 
     */
    public boolean checkColumns(){
        boolean full = true;
        for(int i=1; i<board_row-1; ++i){
            for(int j=1; j<board_row-1; ++j){
                if(Board[i][j] == ' '){
                    full = false;
                    break;
                }
            }
        }
        return full;
    }

    /**
     * If reset button is pressed. It refreshes the board.
     */
    public void clearBoard(){
        for(int r=0; r<board_row; ++r){
            for(int c=0; c<board_col; ++c){
                if(r == 0 || r==board_row-1 || c == 0 || c == board_col-1){
                    Board[r][c] = '#';
                }
                else{
                    Board[r][c] = ' ';
                }            
            }
        }
    }

    /**
     * Returns the 2d char board.
     * 
     * @return tetris Board.
     */
    public char[][] getBoard() {
        return Board;
    }

} // end class Tetris

File: CSE241/HW3/tetris.cpp
#include <iostream>
#include <chrono>
#include <thread>

#include "tetromino.h"
#include "tetris.h"

using namespace std;
using namespace Tetrominos;

namespace TETRIS{

    Tetris::Tetris(){ // No parameter constructor makes a non-usable board
        board_row = 0;
        board_col = 0;
    }

    Tetris::Tetris(int row, int col){ 
        board_row = row + 2;
        board_col = col + 2;

        Board = new char*[board_row];
        for(int i=0; i<board_row; ++i){
            Board[i] = new char[board_row];
        }
        for(int r=0; r<board_row; ++r){
            for(int c=0; c<board_col; ++c){
                if(r == 0 || r == board_row-1 || c == 0 || c == board_col -1){
                    Board[r][c] = '#';
                }
                else{
                    Board[r][c] = ' ';
                }
            }
        }

    }

    Tetris::~Tetris(){ // Destructor
        for(int i=0; i<board_row; ++i){
            delete[] Board[i];
        }
        delete[] Board;
        Board = nullptr;
    }

    void Tetris::Draw() const{ // Prints out the board
        for(int r=0; r<board_row; ++r){
            for(int c=0; c<board_col; ++c){
                cout << Board[r][c];
            }
            cout << endl;
        }
    }

    void Tetris::operator+=(const Tetromino& t){ // Adds tetromino to the top row at mid
        char ** shape = t.getShape();
        int mid_col = board_col / 2;
        for(int r=0; r<t.getRow(); ++r){
            for(int c=0; c<t.getCol(); ++c){
                Board[r+1][mid_col + c - 1] = shape[r][c];
            }
        }
    }

    void Tetris::DeleteAdd(const Tetromino& t){ // Deletes the added tetromino from the top row at mid
        for(int r=0; r<t.getRow(); ++r){
            for(int c=0; c<t.getCol(); ++c){
                Board[r+1][board_col/2 + c - 1] = ' ';
            }
        }
    }

    void Tetris::Animate(Tetromino &t,char move_d, int move_c,char rotate,int rot_c){
        int count = 0;
        int bottom = board_row - 1;
        for(int i=0; i<rot_c; ++i){ // Rotate the shape of the tetromino rotation count times
            t.rotate(rotate);
        }
        char **shape = t.getShape();
        cout << "Shape of tetromino:" << endl;
        t.print();
        if(move_d == 'l'){ // If the given move direction is to the left make move count minus.
            move_c *=-1;
        }
        while(count <= bottom){
            this->Draw();
            this->DeleteAdd(t);
            if(move_d == 'r'){ // If the move direction is to the right and out of border of the board
                if(move_c + board_col / 2 > board_col-2){
                    cout << "You have exceed the boards border" << endl;
                    break;
                }
            }
            else if(move_d == 'l'){ // If the move direction is to the left and out of border of the board
                if(board_col / 2 + move_c < 2){
                    cout << "You have exceed the boards border" << endl;
                    break;
                }
            }
            
            

            for(int i=1; i<t.getCol()+1; ++i){ // Collision check
                if(Board[i+count+t.getCol()][board_col/2+move_c+i] != ' '){
                    return;
                }
            }

            for(int r=0; r<t.getRow(); ++r){ // Erases the previous point
                for(int c=0; c<t.getCol(); ++c){
                    Board[count+1+r][board_col/2+move_c+c-1] = ' ';
                }
            }
            for(int r=0; r<t.getRow(); ++r){ // Puts the shape on the board
                for(int c=0; c<t.getCol(); ++c){
                    Board[count+2+r][board_col/2+move_c+c-1] = shape[r][c];
                }
            }
            this->Sleep();
            count++;
            if(t.getRow() + count >=bottom-1){
                break;
            }
        }
        this->Draw();
    }

    void Tetris::Sleep() const{
        this_thread::sleep_for(chrono::milliseconds(50)); // Sleeps 50 miliseconds
        cout << "\x1b[2J"; // Erases entire screen
        cout << "\x1b[H"; // Move cursor to home position (0,0)
    }




} // end namespace